File: simple_lang_scala/simple_lang_parser/src/main/scala/simple_lang/parser/parser.scala
package simple_lang.parser

import simple_lang.tokenizer._
import scala.util.parsing.combinator._
import scala.util.parsing.input.{NoPosition, Position, Reader}

class ParserException(message: String) extends Exception(message)

object Parser extends Parsers {
    override type Elem = Token
    
    class TokenReader(tokens: Seq[Token]) extends Reader[Token] {
        override def first: Token = tokens.head
        override def atEnd: Boolean = tokens.isEmpty
        override def pos: Position = NoPosition
        override def rest: Reader[Token] = new TokenReader(tokens.tail)
    }
    
    def variable: Parser[VariableToken] = { //we can change everything to lazy vals later as an optimization
        accept("variable", {case id @ VariableToken(name) => id})
    }
    
    def integer: Parser[IntegerLiteralToken] = {
        accept("integer", {case id @ IntegerLiteralToken(value) => id})
    }
    
    def string: Parser[StringLiteralToken] = {
        accept("string", {case id @ StringLiteralToken(value) => id})
    }
    
    def typ: Parser[Type] = {
        IntToken ^^^ IntType |
        BoolToken ^^^ BoolType |
        StringToken ^^^ StringType
    }
    
    def additive_op: Parser[BinOp] = {
        PlusToken ^^^ PlusBinOp |
        MinusToken ^^^ MinusBinOp
    }
    
    def comparison_op: Parser[BinOp] = {
        LessThanToken ^^^ LessThanBinOp |
        GreaterThanToken ^^^ GreaterThanBinOp
    }
    
    def equals_op: Parser[BinOp] = {
        ExactlyEqualsToken ^^^ ExactlyEqualsBinOp |
        NotEqualsToken ^^^ NotEqualsBinOp
    }
    
    def unary_op: Parser[UnaryOp] = {
        NotUnaryToken ^^^ NotUnaryOp |
        MinusUnaryToken ^^^ MinusUnaryOp
    }
    
    def exp_maker(prefix: Exp, list: List[(BinOp ~ Exp)]): Exp = {
        list match {
            case Nil => prefix
            case op ~ exp :: tail => BinOpExp(prefix, op, exp_maker(exp, tail))
        }
    }
    
    def primary_exp: Parser[Exp] = {
        variable ^^ { case value => VariableExp(value.value) } |
        integer ^^ { case value => IntegerLiteralExp(value.value) } |
        string ^^ { case value => StringLiteralExp(value.value) } |
        TrueBooleanLiteralToken ^^^ TrueLiteralExp |
        FalseBooleanLiteralToken ^^^ FalseLiteralExp
    }
    
    def additive_exp: Parser[Exp] = {
        primary_exp ~ rep(additive_op ~ primary_exp) ^^ { case exp ~ list => exp_maker(exp, list) }
    }
    
    def comparison_exp: Parser[Exp] = {
        additive_exp ~ rep(comparison_op ~ additive_exp) ^^ { case exp ~ list => exp_maker(exp, list) }
    }
    
    def equals_exp: Parser[Exp] = {
        comparison_exp ~ rep(equals_op ~ comparison_exp) ^^ { case exp ~ list => exp_maker(exp, list) }
    }
    
    def unary_exp: Parser[Exp] = {
        unary_op ~ equals_exp ^^ { case op ~ exp => UnOpExp(op, exp) } |
        equals_exp ^^ { case exp => exp }
    }
    
    def exp: Parser[Exp] = {
        unary_exp ^^ { case exp => exp }      
    }

    def stmt: Parser[Stmt] = {
        typ ~ variable ~ AssignmentToken ~ exp ~ SemicolonToken ^^ { case annotation ~ var_name ~ _ ~ exp ~ _ => VariableDeclarationStmt(annotation, VariableExp(var_name.value), exp) } |
        variable ~ AssignmentToken ~ exp ~ SemicolonToken ^^ { case var_name ~ _ ~ exp ~ _ => AssignmentStmt(VariableExp(var_name.value), exp) } |
        IfToken ~ LeftParenToken ~ exp ~ RightParenToken ~ stmt ~ ElseToken ~ stmt ^^ { 
            case _ ~ _ ~ exp ~ _ ~ stmt1 ~ _ ~ stmt2 => IfElseStmt(exp, stmt1, stmt2)
        } |
        WhileToken ~ LeftParenToken ~ exp ~ RightParenToken ~ stmt ^^ { case _ ~ _ ~ exp ~ _ ~ stmt => WhileStmt(exp, stmt) } |
        PrintToken ~ LeftParenToken ~ exp ~ RightParenToken ~ SemicolonToken ^^ { case _ ~ _ ~ exp ~ _  ~ _ => PrintStmt(exp) } |
        LeftCurlyBraceToken ~ rep(stmt) ~ RightCurlyBraceToken ^^ { case _ ~ stmts ~ _ => BlockStmt(stmts) }
    }
    
    def program: Parser[Program] = {
        rep(stmt) ^^ { case stmts => Program(stmts) }
    }
    
    def phrase_program: Parser[Program] = {
        phrase(program)
    }
    
    def apply(tokens: Seq[Token]): Program = {
        val reader = new TokenReader(tokens)
        phrase_program(reader) match {
            case Success(result, _) => result
            case Failure(message, _) => throw new ParserException("Parsing Failure")
            case Error(message, _) => throw new ParserException("Parsing Error")
        }
    }
}

File: simple_lang_java/simple-lang-parser-mvn/simple-lang-parser/src/main/java/com/simple_lang/parser/Parser.java
package com.simple_lang.parser;

import simple_lang.tokenizer.*;

import java.util.List;
import java.util.ArrayList;

public class Parser {
    private final List<Token> tokens;
    
    public Parser(final List<Token> tokens) {
        this.tokens = tokens;
    }
    
    public Token getToken(final int position) throws ParseException {
        if (position >= 0 && position < tokens.size()) {
            return tokens.get(position);
        } else {
            throw new ParseException("Invalid token position: " + position);
        }
    }
    
    public void assertTokenHereIs(final int position, final Token expected) throws ParseException {
        final Token received = getToken(position);
        if (!expected.equals(received)) {
            throw new ParseException("expected: " + expected + " but received: " + received);
        }
    }
    
    public ParseResult<Type> parseType(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof IntToken) {
            return new ParseResult<Type>(new IntType(), position + 1);
        } else if (token instanceof BoolToken) {
            return new ParseResult<Type>(new BoolType(), position + 1);
        } else if (token instanceof StringToken) {
            return new ParseResult<Type>(new StringType(), position + 1);
        } else {
            throw new ParseException("Could not parse type.");
        }
    }

    public ParseResult<Exp> parsePrimaryExp(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof VariableToken) {
            final VariableToken varToken = (VariableToken)token;
            return new ParseResult<Exp>(new VariableExp(varToken.name), position + 1);
        } else if (token instanceof IntegerLiteralToken) {
            final IntegerLiteralToken intToken = (IntegerLiteralToken)token;
            return new ParseResult<Exp>(new IntegerExp(intToken.value), position + 1);
        } else if (token instanceof TrueBooleanLiteralToken) {
            return new ParseResult<Exp>(new TrueExp(), position + 1);
        } else if (token instanceof FalseBooleanLiteralToken) {
            return new ParseResult<Exp>(new FalseExp(), position + 1);
        } else if (token instanceof StringLiteralToken) {
            final StringLiteralToken stringToken = (StringLiteralToken)token;
            return new ParseResult<Exp>(new StringExp(stringToken.value), position + 1);
        } else {
            throw new ParseException("Failed parsing primary expression.");
        }
    }
    
    public ParseResult<BinOp> parseAdditiveOp(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof PlusToken) {
            return new ParseResult<BinOp>(new PlusBinOp(), position + 1);
        } else if (token instanceof MinusToken) {
            return new ParseResult<BinOp>(new MinusBinOp(), position + 1);
        } else {
            throw new ParseException("Failed when parsing additive op");
        }
    }
    
    public ParseResult<Exp> parseAdditiveExp(final int position) throws ParseException {
        ParseResult<Exp> current = parsePrimaryExp(position);
        final Token token = getToken(current.position);
        if (token instanceof PlusToken || token instanceof MinusToken) {
            ParseResult<BinOp> op = parseAdditiveOp(current.position);
            ParseResult<Exp> rhs = parseExp(op.position);
            return new ParseResult<Exp>(new BinOpExp(current.result, op.result, rhs.result), rhs.position);
        } else {
            return current;
        }
    }
    
    public ParseResult<BinOp> parseComparisonOp(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof LessThanToken) {
            return new ParseResult<BinOp>(new LessThanBinOp(), position + 1);
        } else if (token instanceof GreaterThanToken) {
            return new ParseResult<BinOp>(new GreaterThanBinOp(), position + 1);
        } else {
            throw new ParseException("Failed when parsing comparison op.");
        }
    }
    
    public ParseResult<Exp> parseComparisonExp(final int position) throws ParseException {
        ParseResult<Exp> current = parseAdditiveExp(position);
        final Token token = getToken(current.position);
        if (token instanceof LessThanToken || token instanceof GreaterThanToken) {
            ParseResult<BinOp> op = parseComparisonOp(current.position);
            ParseResult<Exp> rhs = parseExp(op.position);
            return new ParseResult<Exp>(new BinOpExp(current.result, op.result, rhs.result), rhs.position);
        } else {
            return current;
        }
    }
    
    public ParseResult<BinOp> parseEqualsOp(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof ExactlyEqualsToken) {
            return new ParseResult<BinOp>(new ExactlyEqualsBinOp(), position + 1);
        } else if (token instanceof NotEqualsToken) {
            return new ParseResult<BinOp>(new NotEqualsBinOp(), position + 1);
        } else {
            throw new ParseException("Failed parsing equals op.");
        }
    }
    
    public ParseResult<Exp> parseEqualsExp(final int position) throws ParseException {
        ParseResult<Exp> current = parseComparisonExp(position);
        final Token token = getToken(current.position);
        if (token instanceof ExactlyEqualsToken || token instanceof NotEqualsToken) {
            ParseResult<BinOp> op = parseEqualsOp(current.position);
            ParseResult<Exp> rhs = parseExp(op.position);
            return new ParseResult<Exp>(new BinOpExp(current.result, op.result, rhs.result), rhs.position);
        } else {
            return current;
        }
    }
    
    public ParseResult<UnaryOp> parseUnaryOp(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof NotUnaryToken) {
            return new ParseResult<UnaryOp>(new NotUnaryOp(), position + 1);
        } else if (token instanceof MinusUnaryToken) {
            return new ParseResult<UnaryOp>(new NegationUnaryOp(), position + 1);
        } else {
            throw new ParseException("Failed parsing unary op.");
        }
    }
    
    public ParseResult<Exp> parseUnaryExp(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof NotUnaryToken || token instanceof MinusUnaryToken) {
            final ParseResult<UnaryOp> op = parseUnaryOp(position);
            final ParseResult<Exp> equalsExp = parseEqualsExp(op.position);
            return new ParseResult<Exp>(new UnaryExp(op.result, equalsExp.result), equalsExp.position);
        } else {
            return (parseEqualsExp(position));
        }
    }
    
    public ParseResult<Exp> parseExp(final int position) throws ParseException {
        return (parseUnaryExp(position));
        
    }
    
    public ParseResult<Stmt> parseVarDecStmt(final int position) throws ParseException {
        final ParseResult<Type> type = parseType(position);
        final Token token1 = getToken(type.position);
        if (token1 instanceof VariableToken) {
            final String varName = ((VariableToken)token1).name;
            assertTokenHereIs(type.position + 1, (new AssignmentToken()));
            final ParseResult<Exp> exp = parseExp(type.position + 2);
            assertTokenHereIs(exp.position, (new SemicolonToken()));
            return new ParseResult<Stmt>(new VarDecStmt(type.result, new VariableExp(varName), exp.result), exp.position + 1);
        } else {
            throw new ParseException("Could not parse variable declaration stmt");
        }
    }
    
    public ParseResult<Stmt> parseAssignmentStmt(final int position) throws ParseException {
        final Token token = getToken(position);
        final VariableToken varToken = (VariableToken)token;
        assertTokenHereIs(position + 1, new AssignmentToken());
        final ParseResult<Exp> exp = parseExp(position + 2);
        assertTokenHereIs(exp.position, new SemicolonToken());
        return new ParseResult<Stmt>(new AssignmentStmt(new VariableExp(varToken.name), exp.result), exp.position + 1);
    }
    
    public ParseResult<Stmt> parseIfElseStmt(final int position) throws ParseException {
        final ParseResult<Exp> predicate = parseExp(position);
        assertTokenHereIs(predicate.position, new RightParenToken());
        final ParseResult<Stmt> trueBranch = parseStmt(predicate.position + 1);
        assertTokenHereIs(trueBranch.position, new ElseToken());
        final ParseResult<Stmt> falseBranch = parseStmt(trueBranch.position + 1);
        return new ParseResult<Stmt>(new IfElseStmt(predicate.result, trueBranch.result, falseBranch.result), falseBranch.position);
    }
    
    public ParseResult<Stmt> parseWhileStmt(final int position) throws ParseException {
        final ParseResult<Exp> predicate = parseExp(position);
        assertTokenHereIs(predicate.position, new RightParenToken());
        final ParseResult<Stmt> body = parseStmt(predicate.position + 1);
        return new ParseResult<Stmt>(new WhileStmt(predicate.result, body.result), body.position);
    }
    
    public ParseResult<Stmt> parsePrintStmt(final int position) throws ParseException {
        final ParseResult<Exp> exp = parseExp(position);
        assertTokenHereIs(exp.position, new RightParenToken());
        assertTokenHereIs(exp.position + 1, new SemicolonToken());
        return new ParseResult<Stmt>(new PrintStmt(exp.result), exp.position + 2);
    }
    
    public ParseResult<Stmt> parseBlockStmt(final int position) throws ParseException {
        ParseResult<Stmt> stmt;   //need this so i have access to positional info outside of loop
        Token nextToken;    //same as above
        List<Stmt> stmts = new ArrayList<>();
        final Token token = getToken(position);
        if (token instanceof RightCurlyBraceToken) {
            return new ParseResult<Stmt>(new BlockStmt(stmts), position + 1);
        } else {
            do {
                stmt = parseStmt(position + 1);
                stmts.add(stmt.result);
                nextToken = getToken(stmt.position);
            } while (!(nextToken instanceof RightCurlyBraceToken));
            return new ParseResult<Stmt>(new BlockStmt(stmts), stmt.position + 1);
        }
    }
    
    public ParseResult<Stmt> parseStmt(final int position) throws ParseException {
        final Token token = getToken(position);
        if (token instanceof IntToken || token instanceof BoolToken || token instanceof StringToken) {
            return parseVarDecStmt(position);
        } else if (token instanceof VariableToken) {
            return parseAssignmentStmt(position);
        } else if (token instanceof IfToken) {
            return parseIfElseStmt(position + 2); //skipping the if and the left paren
        } else if (token instanceof WhileToken) {
            return parseWhileStmt(position + 2); //skipping the while and the left paren
        } else if (token instanceof PrintToken) {
            return parsePrintStmt(position + 2); //skipping the print and the left paren
        } else if (token instanceof LeftCurlyBraceToken) {
            return parseBlockStmt(position + 1); //skipping the left curly brace
        } else {
            throw new ParseException("Failed when parsing stmt");
        }
    }
    
    public ParseResult<Program> parseProgram(final int position) throws ParseException {
        List<Stmt> stmts = new ArrayList<Stmt>();
        boolean shouldRunStmts = true;
        int newPosition = position;
        while(shouldRunStmts) {
            try {
                final ParseResult<Stmt> stmt = parseStmt(newPosition);
                //System.out.println("Just parsed this stmt: " + stmt.result.toString());
                stmts.add(stmt.result);
                //System.out.println("The current list of stmts is: " + stmts.toString());
                newPosition = stmt.position;
            } catch (final ParseException e) {
                shouldRunStmts = false;
            }
        }
        return new ParseResult<Program>(new Program(stmts), newPosition);
    }
    
    public Program parseProgram() throws ParseException {
        final ParseResult<Program> program = parseProgram(0);
        if (program.position == tokens.size()) {
            //System.out.println("This is the program I'm gonna return: " + program.result.toString());
            return program.result;
        } else {
            throw new ParseException("Did not parse all tokens, current parse: " + program.result.toString());
        }
    }
}